src/transport/SecureSessionMgr.cpp

```diff
@@ -349,6 +372,44 @@ void SecureSessionMgr::OnMessageReceived(const PacketHeader & packetHeader, cons
         ExitNow(err = CHIP_ERROR_KEY_NOT_FOUND_FROM_PEER);
     }
 
+    if (packetHeader.GetFlags().Has(Header::FlagValues::kSecure))
+    {
+        // Verify message counter
+        if (packetHeader.GetFlags().Has(Header::FlagValues::kSecureSessionControlMessage))
+        {
+            // TODO: control message counter is not implemented yet
+        }
+        else
+        {
+            // TODO: "initial Session Establishment bootstrap" is under specified, use message counter sync protocol for both group
+            // and unicast messages
+            if (!mMessageCounterManager->IsSyncCompleted(state->GetSessionMessageCounter().GetPeerMessageCounter()))
+            {
+                // Queue and start message sync procedure
+                err = mMessageCounterManager->QueueReceivedMessageAndStartSync(
+                    { state->GetPeerNodeId(), state->GetPeerKeyID(), state->GetAdminId() }, state, packetHeader, peerAddress,
+                    std::move(msg));
+
+                if (err != CHIP_NO_ERROR)
+                {
+                    ChipLogError(Inet,
+                                 "Message counter synchronization for received message, failed to "
+                                 "QueueReceivedMessageAndStartSync, err = %d",
+                                 err);
+                }
+
+                return;
+            }
+
+            err = mMessageCounterManager->VerifyCounter(state->GetSessionMessageCounter().GetPeerMessageCounter(), packetHeader);
```

Comment:
This is kind of annoying: the counter is completely derivable from the packet header, but we are passing them both in....  It might almost make more sense to pass in just the message id.

But also, since that just does a trivial call on the counter with the message id and this is the only place that calls VerifyCounter, is the indirection needed?

Reply:
No, the counter is not the counter value, but the counter context. It can not be derived from the packet header.

Reply:
Without need to the Dummy implementation, I can replace them with the direct call.